wallet/src/main/scala/org/bitcoins/wallet/Wallet.scala

```diff
@@ -30,6 +38,197 @@ sealed abstract class Wallet extends LockedWallet with UnlockedWalletApi {
+                   BIP32Node(index = 0, hardened = false))).get,
+        // Refund Key
+        SegWitHDPath(
+          prefix ++
+            Vector(BIP32Node(index = 1, hardened = false),
+                   BIP32Node(index = 0, hardened = false))).get,
+        // Contract Execution Transaction (CET) Keys
+        SegWitHDPath(
+          prefix ++
+            Vector(BIP32Node(index = 1, hardened = false),
+                   BIP32Node(index = 1, hardened = false))).get,
+        SegWitHDPath(
+          prefix ++
+            Vector(BIP32Node(index = 1, hardened = false),
+                   BIP32Node(index = 2, hardened = false))).get,
+        // Closing Keys
+        SegWitHDPath(
+          prefix ++
+            Vector(BIP32Node(index = 2, hardened = false),
+                   BIP32Node(index = 0, hardened = false))).get,
+        SegWitHDPath(
+          prefix ++
+            Vector(BIP32Node(index = 2, hardened = false),
+                   BIP32Node(index = 1, hardened = false))).get,
+        SegWitHDPath(
+          prefix ++
+            Vector(BIP32Node(index = 2, hardened = false),
+                   BIP32Node(index = 2, hardened = false))).get
+      )
+
+      paths.map { addrPath =>
+        val pubkey = keyManager.getRootXPub.deriveChildPubKey(addrPath) match {
+          case Failure(exception) => throw exception
+          case Success(value)     => value.key
+        }
+
+        AddressDbHelper
+          .getSegwitAddress(pubkey, addrPath, networkParameters)
+      }
+    }
+
+    addressDAO.createAll(addressDbs)
+  }
+
+  /** This is needed to fill in values that we do not have yet */
+  private val dummyPartialSig: PartialSignature =
+    PartialSignature(ECPublicKey.freshPublicKey, DummyECDigitalSignature)
+
+  override def createDLCOffer(
+      amount: Bitcoins,
+      oracleInfo: OracleInfo,
+      contractInfoVec: Vector[Sha256DigestBE],
+      feeRateOpt: Option[SatoshisPerVirtualByte],
+      locktime: UInt32,
+      refundLocktime: UInt32): Future[DLCOffer] = {
+    logger.debug("Calculating relevant wallet data for DLC Offer")
+
+    val contractInfo =
+      ContractInfo(
+        (Map.newBuilder ++= contractInfoVec.map(sha => (sha, amount.satoshis)))
+          .result())
+
+    val timeouts = DLCTimeouts(5,
+                               BlockStamp(locktime.toInt),
+                               BlockStamp(refundLocktime.toInt))
+
+    val feeRate = feeRateOpt.getOrElse(SatoshisPerVirtualByte.one)
+
+    val eventId = DLCMessage.calcEventId(oracleInfo, contractInfo, timeouts)
+
+    logger.debug("Checking if offer has already been made")
+    dlcDAO
+      .findByEventId(eventId)
+      .zip(getDefaultAccountForType(AddressType.SegWit))
+      .flatMap {
+        case (dlcDbOpt, account) =>
+          dlcDbOpt match {
+            case Some(dlcDb) =>
+              logger.debug("DLC Offer has already been made, returning offer")
+              listUtxos(dlcDb.fundingInputs).map { spendingInfoDbs =>
```

Comment:
would be much more readable if there were some private sub-defs in here :) or at the very least intermediate vals